r/aws_default_route_table: Refactor acceptance tests in preparation for future fixes/enhancements

ewbankkit opened the pull request:
<!--- See what makes a good Pull Request at : https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/CONTRIBUTING.md --->

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

<!--- Thank you for keeping this note for the community --->

<!--- If your PR fully resolves and should automatically close the linked issue, use Closes. Otherwise, use Relates --->
Refactoring `aws_default_route_table` resource acceptance tests in preparation for future fixes/enhancements.
Relates #13527.

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):
<!--
If change is not user facing, just write "NONE" in the release-note block below.
-->

```release-note
NONE
```

Output from acceptance testing:

<!--
Replace TestAccXXX with a pattern that matches the tests affected by this PR.

For more information on the `-run` flag, see the `go test` documentation at https://tip.golang.org/cmd/go/#hdr-Testing_flags.
-->
```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSDefaultRouteTable_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -count 1 -parallel 2 -run=TestAccAWSDefaultRouteTable_ -timeout 120m
=== RUN   TestAccAWSDefaultRouteTable_basic
=== PAUSE TestAccAWSDefaultRouteTable_basic
=== RUN   TestAccAWSDefaultRouteTable_vpcDisappears
=== PAUSE TestAccAWSDefaultRouteTable_vpcDisappears
=== RUN   TestAccAWSDefaultRouteTable_Route_ConfigMode
=== PAUSE TestAccAWSDefaultRouteTable_Route_ConfigMode
=== RUN   TestAccAWSDefaultRouteTable_swap
=== PAUSE TestAccAWSDefaultRouteTable_swap
=== RUN   TestAccAWSDefaultRouteTable_IPv4_To_TransitGateway
=== PAUSE TestAccAWSDefaultRouteTable_IPv4_To_TransitGateway
=== RUN   TestAccAWSDefaultRouteTable_VpcEndpointAssociation
=== PAUSE TestAccAWSDefaultRouteTable_VpcEndpointAssociation
=== RUN   TestAccAWSDefaultRouteTable_tags
=== PAUSE TestAccAWSDefaultRouteTable_tags
=== RUN   TestAccAWSDefaultRouteTable_ConditionalCidrBlock
=== PAUSE TestAccAWSDefaultRouteTable_ConditionalCidrBlock
=== CONT  TestAccAWSDefaultRouteTable_basic
=== CONT  TestAccAWSDefaultRouteTable_IPv4_To_TransitGateway
--- PASS: TestAccAWSDefaultRouteTable_basic (53.09s)
=== CONT  TestAccAWSDefaultRouteTable_ConditionalCidrBlock
--- PASS: TestAccAWSDefaultRouteTable_ConditionalCidrBlock (78.73s)
=== CONT  TestAccAWSDefaultRouteTable_tags
--- PASS: TestAccAWSDefaultRouteTable_tags (82.10s)
=== CONT  TestAccAWSDefaultRouteTable_VpcEndpointAssociation
--- PASS: TestAccAWSDefaultRouteTable_VpcEndpointAssociation (58.35s)
=== CONT  TestAccAWSDefaultRouteTable_Route_ConfigMode
--- PASS: TestAccAWSDefaultRouteTable_IPv4_To_TransitGateway (318.88s)
=== CONT  TestAccAWSDefaultRouteTable_swap
--- PASS: TestAccAWSDefaultRouteTable_Route_ConfigMode (99.49s)
=== CONT  TestAccAWSDefaultRouteTable_vpcDisappears
--- PASS: TestAccAWSDefaultRouteTable_vpcDisappears (24.11s)
--- PASS: TestAccAWSDefaultRouteTable_swap (106.31s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	425.257s
```

ewbankkit commented on aws/resource_aws_route_table_test.go:
```diff
@@ -555,6 +555,16 @@ func TestAccAWSRouteTable_ConditionalCidrBlock(t *testing.T) {
 	})
 }
 
+func testAccCheckAWSRouteTableNumberOfRoutes(routeTable *ec2.RouteTable, n int) resource.TestCheckFunc {
```
Shared with https://github.com/terraform-providers/terraform-provider-aws/pull/14013.

bflad replied:
We should try to move away from API object `TestCheckFunc` like these in preference of state value and ImportStateVerify checks -- please see: https://github.com/hashicorp/terraform-provider-aws/issues/16435

ewbankkit replied:
In general I completely agree that those kind of checkers aren't desirable, in this case I am verifying that the number of routes returned by the API is _NOT_ equal to the number of routes stored in state:

```go
testAccCheckAWSRouteTableNumberOfRoutes(&routeTable, 2),
resource.TestCheckResourceAttr(resourceName, "route.#", "1"),
```

effectively testing the validity of this code:

https://github.com/hashicorp/terraform-provider-aws/blob/346cac93aba3fd8f87b2de981c9213b74c44ad86/aws/resource_aws_route_table.go#L211-L224

ewbankkit replied:
Happy to remove if you think this is overkill.

ewbankkit commented:
Rebased to fix merge conflicts.
Re-ran acceptance tests:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSDefaultRouteTable_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 2 -run=TestAccAWSDefaultRouteTable_ -timeout 120m
=== RUN   TestAccAWSDefaultRouteTable_basic
=== PAUSE TestAccAWSDefaultRouteTable_basic
=== RUN   TestAccAWSDefaultRouteTable_vpcDisappears
=== PAUSE TestAccAWSDefaultRouteTable_vpcDisappears
=== RUN   TestAccAWSDefaultRouteTable_Route_ConfigMode
=== PAUSE TestAccAWSDefaultRouteTable_Route_ConfigMode
=== RUN   TestAccAWSDefaultRouteTable_swap
=== PAUSE TestAccAWSDefaultRouteTable_swap
=== RUN   TestAccAWSDefaultRouteTable_IPv4_To_TransitGateway
=== PAUSE TestAccAWSDefaultRouteTable_IPv4_To_TransitGateway
=== RUN   TestAccAWSDefaultRouteTable_VpcEndpointAssociation
=== PAUSE TestAccAWSDefaultRouteTable_VpcEndpointAssociation
=== RUN   TestAccAWSDefaultRouteTable_tags
=== PAUSE TestAccAWSDefaultRouteTable_tags
=== RUN   TestAccAWSDefaultRouteTable_ConditionalCidrBlock
=== PAUSE TestAccAWSDefaultRouteTable_ConditionalCidrBlock
=== CONT  TestAccAWSDefaultRouteTable_basic
=== CONT  TestAccAWSDefaultRouteTable_VpcEndpointAssociation
--- PASS: TestAccAWSDefaultRouteTable_basic (33.27s)
=== CONT  TestAccAWSDefaultRouteTable_IPv4_To_TransitGateway
--- PASS: TestAccAWSDefaultRouteTable_VpcEndpointAssociation (45.59s)
=== CONT  TestAccAWSDefaultRouteTable_swap
--- PASS: TestAccAWSDefaultRouteTable_swap (80.37s)
=== CONT  TestAccAWSDefaultRouteTable_Route_ConfigMode
--- PASS: TestAccAWSDefaultRouteTable_Route_ConfigMode (70.97s)
=== CONT  TestAccAWSDefaultRouteTable_vpcDisappears
--- PASS: TestAccAWSDefaultRouteTable_vpcDisappears (14.60s)
=== CONT  TestAccAWSDefaultRouteTable_ConditionalCidrBlock
--- PASS: TestAccAWSDefaultRouteTable_ConditionalCidrBlock (57.40s)
=== CONT  TestAccAWSDefaultRouteTable_tags
--- PASS: TestAccAWSDefaultRouteTable_tags (54.05s)
--- PASS: TestAccAWSDefaultRouteTable_IPv4_To_TransitGateway (312.77s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	346.094s
```

ewbankkit commented:
##### TODO

Merge https://github.com/terraform-providers/terraform-provider-aws/pull/14864 and revisit.

ewbankkit commented:
Rebased and fixed merge conflict.
Re-ran acceptance tests:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSDefaultRouteTable_' ACCTEST_PARALLELISM=2
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 2 -run=TestAccAWSDefaultRouteTable_ -timeout 120m
=== RUN   TestAccAWSDefaultRouteTable_basic
=== PAUSE TestAccAWSDefaultRouteTable_basic
=== RUN   TestAccAWSDefaultRouteTable_vpcDisappears
=== PAUSE TestAccAWSDefaultRouteTable_vpcDisappears
=== RUN   TestAccAWSDefaultRouteTable_Route_ConfigMode
=== PAUSE TestAccAWSDefaultRouteTable_Route_ConfigMode
=== RUN   TestAccAWSDefaultRouteTable_swap
=== PAUSE TestAccAWSDefaultRouteTable_swap
=== RUN   TestAccAWSDefaultRouteTable_IPv4_To_TransitGateway
=== PAUSE TestAccAWSDefaultRouteTable_IPv4_To_TransitGateway
=== RUN   TestAccAWSDefaultRouteTable_VpcEndpointAssociation
=== PAUSE TestAccAWSDefaultRouteTable_VpcEndpointAssociation
=== RUN   TestAccAWSDefaultRouteTable_tags
=== PAUSE TestAccAWSDefaultRouteTable_tags
=== RUN   TestAccAWSDefaultRouteTable_ConditionalCidrBlock
=== PAUSE TestAccAWSDefaultRouteTable_ConditionalCidrBlock
=== CONT  TestAccAWSDefaultRouteTable_basic
=== CONT  TestAccAWSDefaultRouteTable_VpcEndpointAssociation
--- PASS: TestAccAWSDefaultRouteTable_basic (36.82s)
=== CONT  TestAccAWSDefaultRouteTable_ConditionalCidrBlock
--- PASS: TestAccAWSDefaultRouteTable_VpcEndpointAssociation (49.81s)
=== CONT  TestAccAWSDefaultRouteTable_tags
--- PASS: TestAccAWSDefaultRouteTable_ConditionalCidrBlock (64.67s)
=== CONT  TestAccAWSDefaultRouteTable_swap
--- PASS: TestAccAWSDefaultRouteTable_tags (65.13s)
=== CONT  TestAccAWSDefaultRouteTable_IPv4_To_TransitGateway
=== CONT  TestAccAWSDefaultRouteTable_swap
    resource_aws_default_route_table_test.go:157: [INFO] Got non-empty plan, as expected
    resource_aws_default_route_table_test.go:157: [INFO] Got non-empty plan, as expected
--- PASS: TestAccAWSDefaultRouteTable_swap (103.82s)
=== CONT  TestAccAWSDefaultRouteTable_Route_ConfigMode
--- PASS: TestAccAWSDefaultRouteTable_Route_ConfigMode (96.16s)
=== CONT  TestAccAWSDefaultRouteTable_vpcDisappears
    resource_aws_default_route_table_test.go:65: [INFO] Got non-empty plan, as expected
--- PASS: TestAccAWSDefaultRouteTable_vpcDisappears (17.55s)
--- PASS: TestAccAWSDefaultRouteTable_IPv4_To_TransitGateway (323.51s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	438.504s
```

DrFaust92 reviewed (approved):
LGTM

```
--- PASS: TestAccAWSDefaultRouteTable_basic (78.56s)
--- PASS: TestAccAWSDefaultRouteTable_VpcEndpointAssociation (92.45s)
--- PASS: TestAccAWSDefaultRouteTable_ConditionalCidrBlock (126.06s)
--- PASS: TestAccAWSDefaultRouteTable_tags (142.06s)
--- PASS: TestAccAWSDefaultRouteTable_vpcDisappears (38.37s)
--- PASS: TestAccAWSDefaultRouteTable_swap (221.18s)
--- PASS: TestAccAWSDefaultRouteTable_Route_ConfigMode (205.23s)
--- PASS: TestAccAWSDefaultRouteTable_IPv4_To_TransitGateway (444.63s)
```

ewbankkit commented:
When I run the full set of acceptance tests I get:

```
TF_ACC=1 go test ./aws -v -count 1 -parallel 2 -run=TestAccAWSDefaultRouteTable_ -timeout 120m
=== RUN   TestAccAWSDefaultRouteTable_basic
=== PAUSE TestAccAWSDefaultRouteTable_basic
=== RUN   TestAccAWSDefaultRouteTable_vpcDisappears
=== PAUSE TestAccAWSDefaultRouteTable_vpcDisappears
=== RUN   TestAccAWSDefaultRouteTable_Route_ConfigMode
=== PAUSE TestAccAWSDefaultRouteTable_Route_ConfigMode
=== RUN   TestAccAWSDefaultRouteTable_swap
=== PAUSE TestAccAWSDefaultRouteTable_swap
=== RUN   TestAccAWSDefaultRouteTable_IPv4_To_TransitGateway
=== PAUSE TestAccAWSDefaultRouteTable_IPv4_To_TransitGateway
=== RUN   TestAccAWSDefaultRouteTable_IPv4_To_VpcEndpoint
=== PAUSE TestAccAWSDefaultRouteTable_IPv4_To_VpcEndpoint
=== RUN   TestAccAWSDefaultRouteTable_VpcEndpointAssociation
=== PAUSE TestAccAWSDefaultRouteTable_VpcEndpointAssociation
=== RUN   TestAccAWSDefaultRouteTable_tags
=== PAUSE TestAccAWSDefaultRouteTable_tags
=== RUN   TestAccAWSDefaultRouteTable_ConditionalCidrBlock
=== PAUSE TestAccAWSDefaultRouteTable_ConditionalCidrBlock
=== CONT  TestAccAWSDefaultRouteTable_basic
=== CONT  TestAccAWSDefaultRouteTable_ConditionalCidrBlock
=== CONT  TestAccAWSDefaultRouteTable_tags
--- PASS: TestAccAWSDefaultRouteTable_basic (30.21s)
--- PASS: TestAccAWSDefaultRouteTable_ConditionalCidrBlock (54.98s)
=== CONT  TestAccAWSDefaultRouteTable_VpcEndpointAssociation
--- PASS: TestAccAWSDefaultRouteTable_tags (52.97s)
=== CONT  TestAccAWSDefaultRouteTable_IPv4_To_VpcEndpoint
--- PASS: TestAccAWSDefaultRouteTable_VpcEndpointAssociation (43.83s)
=== CONT  TestAccAWSDefaultRouteTable_IPv4_To_TransitGateway
=== CONT  TestAccAWSDefaultRouteTable_IPv4_To_VpcEndpoint
--- FAIL: TestAccAWSDefaultRouteTable_IPv4_To_VpcEndpoint (117.63s)
=== CONT  TestAccAWSDefaultRouteTable_swap
--- PASS: TestAccAWSDefaultRouteTable_swap (89.43s)
=== CONT  TestAccAWSDefaultRouteTable_Route_ConfigMode
--- PASS: TestAccAWSDefaultRouteTable_Route_ConfigMode (83.12s)
=== CONT  TestAccAWSDefaultRouteTable_vpcDisappears
--- PASS: TestAccAWSDefaultRouteTable_vpcDisappears (14.14s)
--- PASS: TestAccAWSDefaultRouteTable_IPv4_To_TransitGateway (359.22s)
```

with

```
Error: error waiting for VPC Endpoint (vpce-0f812974e9f2a0856) to become available: VPC Endpoint is in a failed state
```

ewbankkit commented:
Rebased and fixed merge conflicts.

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSDefaultRouteTable_' ACCTEST_PARALLELISM=2
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 2 -run=TestAccAWSDefaultRouteTable_ -timeout 120m
=== RUN   TestAccAWSDefaultRouteTable_basic
=== PAUSE TestAccAWSDefaultRouteTable_basic
=== RUN   TestAccAWSDefaultRouteTable_vpcDisappears
=== PAUSE TestAccAWSDefaultRouteTable_vpcDisappears
=== RUN   TestAccAWSDefaultRouteTable_Route_ConfigMode
=== PAUSE TestAccAWSDefaultRouteTable_Route_ConfigMode
=== RUN   TestAccAWSDefaultRouteTable_swap
=== PAUSE TestAccAWSDefaultRouteTable_swap
=== RUN   TestAccAWSDefaultRouteTable_IPv4_To_TransitGateway
=== PAUSE TestAccAWSDefaultRouteTable_IPv4_To_TransitGateway
=== RUN   TestAccAWSDefaultRouteTable_IPv4_To_VpcEndpoint
=== PAUSE TestAccAWSDefaultRouteTable_IPv4_To_VpcEndpoint
=== RUN   TestAccAWSDefaultRouteTable_VpcEndpointAssociation
=== PAUSE TestAccAWSDefaultRouteTable_VpcEndpointAssociation
=== RUN   TestAccAWSDefaultRouteTable_tags
=== PAUSE TestAccAWSDefaultRouteTable_tags
=== RUN   TestAccAWSDefaultRouteTable_ConditionalCidrBlock
=== PAUSE TestAccAWSDefaultRouteTable_ConditionalCidrBlock
=== CONT  TestAccAWSDefaultRouteTable_basic
=== CONT  TestAccAWSDefaultRouteTable_ConditionalCidrBlock
--- PASS: TestAccAWSDefaultRouteTable_basic (30.06s)
=== CONT  TestAccAWSDefaultRouteTable_tags
--- PASS: TestAccAWSDefaultRouteTable_ConditionalCidrBlock (55.68s)
=== CONT  TestAccAWSDefaultRouteTable_VpcEndpointAssociation
--- PASS: TestAccAWSDefaultRouteTable_tags (53.46s)
=== CONT  TestAccAWSDefaultRouteTable_IPv4_To_VpcEndpoint
=== CONT  TestAccAWSDefaultRouteTable_IPv4_To_TransitGateway
--- PASS: TestAccAWSDefaultRouteTable_VpcEndpointAssociation (46.09s)
--- PASS: TestAccAWSDefaultRouteTable_IPv4_To_VpcEndpoint (307.68s)
=== CONT  TestAccAWSDefaultRouteTable_swap
--- PASS: TestAccAWSDefaultRouteTable_IPv4_To_TransitGateway (327.46s)
=== CONT  TestAccAWSDefaultRouteTable_Route_ConfigMode
--- PASS: TestAccAWSDefaultRouteTable_swap (90.87s)
=== CONT  TestAccAWSDefaultRouteTable_vpcDisappears
--- PASS: TestAccAWSDefaultRouteTable_vpcDisappears (14.19s)
--- PASS: TestAccAWSDefaultRouteTable_Route_ConfigMode (84.58s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	513.872s
```

bflad commented on aws/resource_aws_default_route_table_test.go:
```diff
@@ -53,33 +54,37 @@ func TestAccAWSDefaultRouteTable_basic(t *testing.T) {
-func TestAccAWSDefaultRouteTable_disappears_Vpc(t *testing.T) {
-	var routeTable1 ec2.RouteTable
-	var vpc1 ec2.Vpc
+func TestAccAWSDefaultRouteTable_vpcDisappears(t *testing.T) {
```
The more prevalent style for naming these is `_disappears_UPSTREAM`:

```console
$ ag --stats 'func TestAcc.*_disappears_' aws/*_test.go
...
29 matches
25 files contained matches
...
$ ag --stats 'func TestAcc.*_[^_]+Disappears' aws/*_test.go
...
9 matches
9 files contained matches
...
```

I will leave a note for myself to update the contributing guide with the naming recommendation.

```suggestion
func TestAccAWSDefaultRouteTable_disappears_Vpc(t *testing.T) {
```

ewbankkit replied:
Relates:
* https://github.com/hashicorp/terraform-provider-aws/issues/16591
* https://github.com/hashicorp/terraform-provider-aws/pull/16597

ewbankkit commented:
Validated renamed-after-review acceptance test:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSDefaultRouteTable_disappears_Vpc'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSDefaultRouteTable_disappears_Vpc -timeout 120m
=== RUN   TestAccAWSDefaultRouteTable_disappears_Vpc
=== PAUSE TestAccAWSDefaultRouteTable_disappears_Vpc
=== CONT  TestAccAWSDefaultRouteTable_disappears_Vpc
--- PASS: TestAccAWSDefaultRouteTable_disappears_Vpc (15.59s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	15.632s
```

bflad reviewed (approved):
Let's get this in 🚀 

Output from acceptance testing:

```
--- PASS: TestAccAWSDefaultRouteTable_basic (79.29s)
--- PASS: TestAccAWSDefaultRouteTable_ConditionalCidrBlock (108.38s)
--- PASS: TestAccAWSDefaultRouteTable_disappears_Vpc (33.90s)
--- PASS: TestAccAWSDefaultRouteTable_IPv4_To_TransitGateway (375.62s)
--- PASS: TestAccAWSDefaultRouteTable_IPv4_To_VpcEndpoint (314.21s)
--- PASS: TestAccAWSDefaultRouteTable_Route_ConfigMode (142.00s)
--- PASS: TestAccAWSDefaultRouteTable_swap (140.28s)
--- PASS: TestAccAWSDefaultRouteTable_tags (116.89s)
--- PASS: TestAccAWSDefaultRouteTable_VpcEndpointAssociation (83.20s)
```
